feat: add support for typescript 3.2

IgorMinar opened the pull request:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-2.html
https://blogs.msdn.microsoft.com/typescript/2018/11/29/announcing-typescript-3-2/

Any application using tsickle for closure compatibility will need to update it's tsickle dependency to 0.34

IgorMinar committed:
feat: add support for typescript 3.2

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-2.html
https://blogs.msdn.microsoft.com/typescript/2018/11/29/announcing-typescript-3-2/

Any application using tsickle for closure compatibility will need to update it's tsickle
dependency to 0.34

mary-poppins commented:
You can preview 0f86dfc at https://pr27536-0f86dfc.ngbuilds.io/.

mary-poppins commented:
You can preview 3ff3b65 at https://pr27536-3ff3b65.ngbuilds.io/.

DevVersion commented on package.json:
```diff
@@ -78,9 +78,9 @@
-    "tsickle": "0.32.1",
-    "tslib": "^1.7.1",
-    "typescript": "~3.1.1",
+    "tsickle": "0.34.0",
```
I have a related PR in #27402 for `@angular/bazel` which also specifies `tsickle` as a dependency. 

Technically it would be good to have `tsickle` a peer dependency for `@angular/bazel`, but that would be a breaking change, so I just updated it to `0.34.0` as well. Should we add these changes here?

IgorMinar replied:
tsickle is intentionally not a dependency or a peer dependency of any of the angular packages.

the reason for this is that each tsickle release supports only one version of typescript (or multiple versions, but almost always a smaller range than angular). Because tsickle is only needed by people that use closure compiler and typescript - which is an advanced usecase - we leave it up to these advanced users to sort out the dependencies manually rather than requiring everyone to install tsickle and having to deal with the delicate nature of tsickle-typescript-closure dependencies.

DevVersion replied:
Makes sense. I see what you mean.

My thought was just that since `tsickle` is _always_ used for `@angular/bazel` to compile an `ng_module` (see [here](https://github.com/angular/angular/blob/master/packages/bazel/src/ngc-wrapped/index.ts#L288)) and the TypeScript range is a bit flexible, two things can happen:

1) TypeScript version range is loose and it's not guaranteed that `tsickle` will work for all allowed TS versions
2) `tsickle` version is too old and doesn't match for any TypeScript version that is allowed as peer dependency.

I think we should just ensure that _if we update the TypeScript range_ for `@angular/bazel` that we also bump the `tsickle` version so that it works, right?

mary-poppins commented:
You can preview 42d7176 at https://pr27536-42d7176.ngbuilds.io/.

mary-poppins commented:
You can preview 8d4ba41 at https://pr27536-8d4ba41.ngbuilds.io/.

mary-poppins commented:
You can preview d784ba0 at https://pr27536-d784ba0.ngbuilds.io/.

mary-poppins commented:
You can preview ff0edd5 at https://pr27536-ff0edd5.ngbuilds.io/.

Commit:
fix(compiler-cli): create LiteralLikeNode for String and Number literal

Typescript 3.2 introduced BigInt type, and consequently the
implementation for checkExpressionWorker() in checkers.ts is refactored.

For NumberLiteral and StringLiteral types, 'text' filed must be present
in the Node type, therefore they must be LiteralLikeNode instead of
Node.

mary-poppins commented:
You can preview f5a6d4c at https://pr27536-f5a6d4c.ngbuilds.io/.

mary-poppins commented:
You can preview 1530a94 at https://pr27536-1530a94.ngbuilds.io/.

alxhub commented on packages/compiler-cli/src/ngcc/src/packages/transformer.ts:
```diff
@@ -8,6 +8,8 @@
 import {existsSync, writeFileSync} from 'fs';
 import {mkdir, mv} from 'shelljs';
+// "unused" import needed by tsc type inference
+import * as ts from 'typescript';
```
I think this is required by the return type of `Transformer.analyzeProgram` being inferred (and the inferred type including TS types). If you make that return type explicit, this statement won't appear unused.

mary-poppins commented:
You can preview a6e6763 at https://pr27536-a6e6763.ngbuilds.io/.

IgorMinar commented:
https://test.corp.google.com/ui#id=OCL:225629145:BASE:225632835:1544839712303:80c11cce

IgorMinar committed:
fix(docs-infra): workaround for broken 'import as'

It's unclear why `import as` results in the aliases to be undefined.

Plain tsc seems to do the right thing and emits the correct code, so it
might be some kind of interaction in @angular/cli or webpack that are
causing the failure.

This should be investigated separately from the tsc update in
angular/angular. See angular/angular-cli#13212

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/packages/transformer.ts:
```diff
@@ -105,7 +107,8 @@ export class Transformer {
 
-  analyzeProgram(reflectionHost: NgccReflectionHost, isCore: boolean, bundle: EntryPointBundle) {
+  analyzeProgram(reflectionHost: NgccReflectionHost, isCore: boolean, bundle: EntryPointBundle):
+      ProgramAnalyses {
```
:+1:

petebacondarwin commented on packages/compiler-cli/src/transformers/program.ts:
```diff
@@ -79,7 +79,7 @@ const MIN_TS_VERSION = '3.1.1';
  * MAX_TS_VERSION is not considered as a supported TypeScript version
  */
-const MAX_TS_VERSION = '3.2.0';
+const MAX_TS_VERSION = '3.3.0';
```
Why not 3.3.2?

mary-poppins commented:
You can preview 91cbd0a at https://pr27536-91cbd0a.ngbuilds.io/.

petebacondarwin committed:
build(docs-infra): update dgeni and dgeni-packages to latest

petebacondarwin commented:
@IgorMinar 3e7c243c6ecfb881fc53791ba4d33fe4dc569065 is not needed for this PR to be green.

IgorMinar commented on aio/src/app/custom-elements/elements-loader.ts:
```diff
@@ -5,9 +5,14 @@ import {
 
+// huh? not sure why, but ts 3.2 emits broken code for: `import { from as fromPromise, ...} from 'rxjs';
+// this const alias works around the problem.
+const fromPromise = from;
```
that's what I'll do for now, I just wanted to get more people to take a look at this because `import as` should work and shouldn't break with a typescript update.

I filed https://github.com/angular/angular-cli/issues/13212 to track this.

IgorMinar committed:
build(docs-infra): disable cli.warnings.typescriptMismatch warnings

cli is not yet officially compatible with typescript 3.2, so we need to disable the version check via:

ng config cli.warnings.typescriptMismatch false

IgorMinar committed:
build(docs-infra): turn on disableTypeScriptVersionCheck in angularCompilerOptions

We need to do this because we don't yet have a released version of angular
with typescript 3.2 support and on CI we test both against the snapshot and
whatever is in aio/yarn.lock.

Once we have the next rc or a stable relase we should be able to remove
this flag.

PS: I also removed the preserveWhitespace:false because that's the default now.

mary-poppins commented:
You can preview 94474cc at https://pr27536-94474cc.ngbuilds.io/.

IgorMinar commented:
http://test/OCL:225629145:BASE:225943315:1545115007312:6f0d4fe4

IgorMinar commented:
caretaker note: pullapprove and travis flaked

zw-slime commented:
Package "ng-zorro-antd" has an incompatible peer dependency to "typescript" (requires "~3.1.1", would install "3.2.2").

I can not update ngrx to 7.0.0,  please help, when can support typescript 3.2?

IgorMinar commented:
angular framework 7.2 has been released on Monday with typescript 3.2
support:
https://github.com/angular/angular/blob/master/CHANGELOG.md#720-2019-01-07

On Tue, Jan 8, 2019 at 7:43 PM zw <notifications@github.com> wrote:

> Package "ng-zorro-antd" has an incompatible peer dependency to
> "typescript" (requires "~3.1.1", would install "3.2.2").
>
> I can not update ngrx to 7.0.0, please help, when can support typescript
> 3.2?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/27536#issuecomment-452561853>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6Ml2VMukiLU4yMdn32KtZ867HHlhks5vBWVPgaJpZM4ZH6Eb>
> .
>
